### ISSUE TITLE
Update up to 4.9 hang at 'waiting on cloud-controller-manager'

### ISSUE THREAD

**Issue by Khuman2** [openshift/okd#996]
**Describe the bug**
The upgrade process hangs. 

oc adm upgrade:
`info: An upgrade is in progress. Working towards 4.9.0-0.okd-2021-11-28-035710: 142 of 744 done (19% complete), waiting on cloud-controller-manager
`

**Version**
```
NAME              STATUS   ROLES    AGE   VERSION                INTERNAL-IP   EXTERNAL-IP   OS-IMAGE           KERNEL-VERSION            CONTAINER-RUNTIME
okd4-compute-01   Ready    worker   74d   v1.21.2+9e8f924-1555   10.4.200.51   <none>        Fedora CoreOS 34   5.14.14-200.fc34.x86_64   cri-o://1.21.4
okd4-compute-02   Ready    worker   74d   v1.21.2+9e8f924-1555   10.4.200.52   <none>        Fedora CoreOS 34   5.14.14-200.fc34.x86_64   cri-o://1.21.4
okd4-master-01    Ready    master   74d   v1.21.2+9e8f924-1555   10.4.200.31   <none>        Fedora CoreOS 34   5.14.14-200.fc34.x86_64   cri-o://1.21.4
okd4-master-02    Ready    master   74d   v1.21.2+9e8f924-1555   10.4.200.32   <none>        Fedora CoreOS 34   5.14.14-200.fc34.x86_64   cri-o://1.21.4
okd4-master-03    Ready    master   74d   v1.21.2+9e8f924-1555   10.4.200.33   <none>        Fedora CoreOS 34   5.14.14-200.fc34.x86_64   cri-o://1.21.4
okd4-master-04    Ready    master   54d   v1.21.2+9e8f924-1555   10.4.200.34   <none>        Fedora CoreOS 34   5.14.14-200.fc34.x86_64   cri-o://1.21.4
okd4-master-05    Ready    master   54d   v1.21.2+9e8f924-1555   10.4.200.35   <none>        Fedora CoreOS 34   5.14.14-200.fc34.x86_64   cri-o://1.21.4
```

**How reproducible**
I don't know

**Link to must gather archive**
https://drive.google.com/file/d/1JIqzEIKhDkY97eHzolzfez27ylyWnATP/view?usp=sharing

**Comment by vrutkovs** [comment]
CVO:
```
task.go:112] error running apply for clusteroperator "cloud-controller-manager" (143 of 744): Cluster operator cloud-controller-manager is updating versions
```
Cloud controller manager operator can't get leader lease:
```
2021-11-29T19:00:45.465242225Z I1129 19:00:45.465100       1 leaderelection.go:248] attempting to acquire leader lease openshift-cloud-controller-manager-operator/cluster-cloud-controller-manager-leader...
```

Seems to be a cloud controller manager operator bug, please open a new issue at https://bugzilla.redhat.com/enter_bug.cgi?component=Cloud%20Compute&product=OpenShift%20Container%20Platform&version=4.9, subcomponent "Cloud Controller manager"

**Comment by bdlink** [comment]
Is there a workaround?  This is happening to me, upgrading from 4.8.0-0.okd-2021-10-24-061736 to 4.9.0-0.okd-2021-12-12-025847 on vsphere UPI.  

clusterversionoperator reports progress, clound controller manager operator is not updating, and cluster version log shows:
error running apply for clusteroperator "cloud-controller-manager" (143 of 744): Cluster operator cloud-controller-manager is updating versions


There is no unsucessful leader lease in the clusterversion operator logs.

**Comment by vrutkovs** [comment]
iiuc this is fixed in 4.10, https://github.com/openshift/cluster-cloud-controller-manager-operator/pull/162 should fix it 4.9.

Since we need a payload update, lets reopen this and close when it lands in stable OKD

**Comment by bdlink** [comment]
I scaled the cluster-cloud-controller-manager pod to 0, restarted the cluster-cloud-controller-manager-operator pod.

The cluster-cloud-controller-manager-operator was able to get the lease and the upgrade is continuing.

**Comment by bdurrow** [comment]
@bdlink, be careful - I used a workaround that sounds like what you did and it left the cluster-cloud-controller-manager deployment (or was it dc, ss or ds).  I am not sure if it is related but I also ended up with broken garbage collection especially noticeable when a job was removed it's completed pod stuck around eventually causing IP exhaustion.  I tracked down that problem to a reference to a webhook in one of my CRDs that was not functioning.

**Comment by kai-uwe-rommel** [comment]
We hit the same problem(s) yesterday when upgrading a cluster from 4.8-2021-10-24 to 4.9-2022-01-14 and we could solve both the cloud-controller-manager and the monitoring problems the same way. That was an upgrade skipping three intermediate versions.
I had upgraded two other clusters during the last months through three of these four upgrades (skipping only 4.9-2021-11-28) and have not hit these problems for these two clusters.

**Comment by kai-uwe-rommel** [comment]
(Meanwhile we upgraded the cluster also to 4.9-2022-01-29 but the GC problems have not been solved.)

**Comment by elmiko** [comment]
just out of curiosity, are you upgrading the feature gate enabled for the external cloud controllers?

i was under the impression that the CCMs were tech preview in 4.9 and required a feature gate to enable. just wondering if i'm missing something here.

**Comment by kai-uwe-rommel** [comment]
I have not heard of any such feature gate nor did I enable anything like this.

**Comment by elmiko** [comment]
interesting, i have a feeling, as Vadim suggested, that this is a bug which was fixed. but do you happen to still have the logs for the cluster-cloud-controller-manager-operator available still?

the CCM/CCCMO were designed to be enabled by a tech preview feature gate for 4.9, so i'm a little stumped about the CCCMO attempting to deploy the CCMs. this is the code where is checks (from the 4.9 branch) https://github.com/openshift/cluster-cloud-controller-manager-operator/blob/release-4.9/pkg/controllers/clusteroperator_controller.go#L62 but perhaps the version in that release payload doesn't have this in place.

**Comment by kai-uwe-rommel** [comment]
No, I can no longer access the logs for this on the cluster.

**Comment by elmiko** [comment]
ack, thanks for bringing up the issue anyway =)

**Comment by JoelSpeed** [comment]
@kai-uwe-rommel Could you possibly share some examples of either stuck pods or garbage collection logs that you've identified that might help us to identify what exactly the issue here is?

**Comment by kai-uwe-rommel** [comment]
It turned out that in this case an orphaned old CRD (runners.apps.gitlab.com) has caused the trouble that we saw after getting through the upgrade with the workaround documented above. This CRD had no relation with CCM.
What remains is that we had the upgrade problem at 19% with CCM at all ... with an OKD version that should already have the fix for it.
